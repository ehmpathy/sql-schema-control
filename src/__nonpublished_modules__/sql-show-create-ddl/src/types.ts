export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<{ rows: any[] }>;
  end: () => Promise<void>;
}
