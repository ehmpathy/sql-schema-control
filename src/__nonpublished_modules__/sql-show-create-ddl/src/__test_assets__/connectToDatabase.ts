import mysql from 'mysql2/promise';
import pg from 'pg';

import { ConnectionConfig, DatabaseLanguage } from '../../../../types';
import { DatabaseConnection } from '../types';

// create the connection to database
const connectionAdapters = {
  [DatabaseLanguage.MYSQL]: async ({
    connectionConfig,
  }: {
    connectionConfig: ConnectionConfig;
  }): Promise<DatabaseConnection> => {
    const connection = await mysql.createConnection({
      host: connectionConfig.host,
      port: connectionConfig.port,
      database: connectionConfig.schema,
      user: connectionConfig.username,
      password: connectionConfig.password,
      multipleStatements: true,
    });
    return {
      query: async ({ sql, values }: { sql: string; values?: any[] }) => {
        const result = await connection.query(sql, values);
        return { rows: result[0] }; // standardize the response
      },
      end: async () => connection.end(),
    };
  },
  [DatabaseLanguage.POSTGRES]: async ({
    connectionConfig,
  }: {
    connectionConfig: ConnectionConfig;
  }): Promise<DatabaseConnection> => {
    const client = new pg.Client({
      host: connectionConfig.host,
      port: connectionConfig.port,
      user: connectionConfig.username,
      password: connectionConfig.password,
      database: connectionConfig.schema,
    });
    await client.connect();
    return {
      query: async ({ sql, values }: { sql: string; values?: (string | number)[] }) => {
        const result = await client.query(sql, values);
        return { rows: result.rows };
      },
      end: () => client.end(),
    };
  },
};

export const connectToDatabase = async ({
  language,
  config,
}: {
  language: DatabaseLanguage;
  config: ConnectionConfig;
}): Promise<DatabaseConnection> => {
  // 1. get the connection adapter for the method
  const getDbConnection = connectionAdapters[language];

  // 2. attempt to connect
  const connection = await getDbConnection({ connectionConfig: config });

  // 3. run a test query to check connection
  await connection.query({ sql: 'SELECT 1' });

  // 4. return the connection
  return connection;
};
