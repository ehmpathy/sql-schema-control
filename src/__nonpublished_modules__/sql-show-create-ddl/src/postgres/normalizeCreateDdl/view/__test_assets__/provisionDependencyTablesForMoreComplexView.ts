import { DatabaseConnection } from '../../../../types';

const dependencyTablesDdl = [
  `
CREATE TABLE home (
  id bigserial NOT NULL,
  uuid uuid NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  name varchar(255) NOT NULL,
  built timestamp with time zone NOT NULL,
  bedrooms int NOT NULL,
  bathrooms int NOT NULL,
  host_ids_hash bytea NOT NULL,
  CONSTRAINT home_pk PRIMARY KEY (id),
  CONSTRAINT home_ux1 UNIQUE (name, host_ids_hash)
);
      `.trim(),
  `
CREATE TABLE home_version (
  id bigserial NOT NULL,
  home_id bigint NOT NULL,
  effective_at timestamp with time zone NOT NULL DEFAULT now(),
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  photo_ids_hash bytea NOT NULL,
  CONSTRAINT home_version_pk PRIMARY KEY (id),
  CONSTRAINT home_version_ux1 UNIQUE (home_id, effective_at, created_at),
  CONSTRAINT home_version_fk0 FOREIGN KEY (home_id) REFERENCES home (id)
);
CREATE INDEX home_version_fk0_ix ON home_version USING btree (home_id);
      `.trim(),
  `
CREATE TABLE home_version_to_photo (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_version_id bigint NOT NULL,
  photo_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_version_to_photo_pk PRIMARY KEY (id),
  CONSTRAINT home_version_to_photo_ux1 UNIQUE (home_version_id, array_order_index),
  CONSTRAINT home_version_to_photo_fk0 FOREIGN KEY (home_version_id) REFERENCES home_version (id)
  -- CONSTRAINT home_version_to_photo_fk1 FOREIGN KEY (photo_id) REFERENCES photo (id)
);
CREATE INDEX home_version_to_photo_fk0_ix ON home_version_to_photo USING btree (home_version_id);
CREATE INDEX home_version_to_photo_fk1_ix ON home_version_to_photo USING btree (photo_id);
      `.trim(),
  `
CREATE TABLE home_to_host (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  host_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_to_host_pk PRIMARY KEY (id),
  CONSTRAINT home_to_host_ux1 UNIQUE (home_id, array_order_index),
  CONSTRAINT home_to_host_fk0 FOREIGN KEY (home_id) REFERENCES home (id)
  -- CONSTRAINT home_to_host_fk1 FOREIGN KEY (host_id) REFERENCES host (id)
);
CREATE INDEX home_to_host_fk0_ix ON home_to_host USING btree (home_id);
CREATE INDEX home_to_host_fk1_ix ON home_to_host USING btree (host_id);
      `.trim(),
  `
CREATE TABLE home_cvp (
  id bigserial NOT NULL,
  updated_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  home_version_id bigint NOT NULL,
  CONSTRAINT home_cvp_pk PRIMARY KEY (id),
  CONSTRAINT home_cvp_ux1 UNIQUE (home_id),
  CONSTRAINT home_cvp_fk0 FOREIGN KEY (home_id) REFERENCES home (id),
  CONSTRAINT home_cvp_fk1 FOREIGN KEY (home_version_id) REFERENCES home_version (id)
);
CREATE INDEX home_cvp_fk0_ix ON home_cvp USING btree (home_id);
CREATE INDEX home_cvp_fk1_ix ON home_cvp USING btree (home_version_id);
      `.trim(),
];
const dependencyTablesNames = [
  'home_cvp',
  'home_to_host',
  'home_version_to_photo',
  'home_version',
  'home',
];
export const provisionDependencyTablesForMoreComplexView = async ({
  dbConnection,
}: {
  dbConnection: DatabaseConnection;
}) => {
  await Promise.all(
    dependencyTablesNames.map((tableName) =>
      dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tableName} CASCADE` }),
    ),
  );
  await Promise.all(
    dependencyTablesDdl.map((ddl) => dbConnection.query({ sql: ddl })),
  );
};
