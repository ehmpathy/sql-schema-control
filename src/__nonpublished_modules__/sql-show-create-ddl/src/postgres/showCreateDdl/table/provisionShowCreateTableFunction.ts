import fs from 'fs';

import { DatabaseConnection } from '../../../types';

/**
 * fortunately for us, postgres does not come with built in `SHOW CREATE TABLE/VIEW/FUNCTION`
 *
 * we have to provide a way to do so manually. on the bright side, postgres does provide an extensive catalog system that we can leverage
 *
 * based on https://stackoverflow.com/questions/2593803/how-to-generate-the-create-table-sql-statement-for-an-existing-table-in-postgr
 */
export const provisionShowCreateTableFunction = async ({ dbConnection }: { dbConnection: DatabaseConnection }) => {
  const sql = await new Promise<string>((resolve, reject) =>
    fs.readFile(`${__dirname}/show_create_table.sql`, (error, data) => {
      if (error) return reject(error);
      return resolve(data.toString());
    }),
  );
  await dbConnection.query({ sql });
};
