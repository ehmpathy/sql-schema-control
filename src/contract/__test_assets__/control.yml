language: mysql
dialect: 5.7
connection: ./connection.config.js
strict: false # strict = true -> display uncontrolled resources
definitions:
  - ./definitions/tables.yml
  - ./definitions/functions.yml
  - ./definitions/procedures.yml
  - ./definitions/init.yml
