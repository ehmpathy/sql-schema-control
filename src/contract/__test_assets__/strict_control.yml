language: mysql
dialect: 5.7
connection: ./connection.config.js
strict: true # strict = true -> display uncontrolled resources
definitions:
  - ./definitions/tables.yml
  - ./definitions/views.yml
  - ./definitions/functions.yml
  - ./definitions/procedures.yml
  - ./definitions/init.yml
