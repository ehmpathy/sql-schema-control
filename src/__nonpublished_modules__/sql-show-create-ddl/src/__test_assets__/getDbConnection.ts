import { DatabaseLanguage } from '../../../../domain';
import { connectToDatabase } from './connectToDatabase';

// the connection config for our integration test dbs, provisioned by docker
export const config = {
  [DatabaseLanguage.MYSQL]: {
    host: 'localhost',
    port: 12821,
    database: 'superimportantdb',
    username: 'root',
    password: 'a-secure-password',
  },
  [DatabaseLanguage.POSTGRES]: {
    host: 'localhost',
    port: 7821,
    database: 'superimportantdb',
    schema: 'public',
    username: 'postgres',
    password: 'a-secure-password',
  },
};

export const getDbConnection = async ({
  language,
}: {
  language: DatabaseLanguage;
}) => connectToDatabase({ language, config: config[language] });
