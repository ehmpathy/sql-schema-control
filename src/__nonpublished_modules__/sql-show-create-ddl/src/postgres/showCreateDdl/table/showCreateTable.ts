import { DatabaseConnection } from '../../../types';
import { provisionShowCreateTableFunction } from './provisionShowCreateTableFunction';

export const showCreateTable = async ({
  dbConnection,
  schema,
  table,
}: {
  dbConnection: DatabaseConnection;
  schema: string;
  table: string;
}) => {
  // provision the show create function if it dne (since nothing like this comes standard w/ postgres)
  const { rows } = await dbConnection.query({
    sql: "select * from information_schema.routines where routine_name = 'show_create_table' ",
  });
  if (!rows.length) {
    // since dne, provision it
    await provisionShowCreateTableFunction({ dbConnection });
  }

  // grab the ddl our function generates
  const result = await dbConnection.query({
    sql: 'SELECT public.show_create_table($1, $2) as ddl',
    values: [schema, table],
  });
  if (!result.rows) throw new Error(`could not find table '${schema}.${table}'`);
  return result.rows[0].ddl;
};
