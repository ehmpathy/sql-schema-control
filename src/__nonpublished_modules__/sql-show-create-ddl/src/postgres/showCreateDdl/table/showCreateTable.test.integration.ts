import { DatabaseLanguage } from '../../../../../../types';
import { getDbConnection } from '../../../__test_assets__/getDbConnection';
import { DatabaseConnection } from '../../../types';
import { provisionShowCreateTableFunction } from './provisionShowCreateTableFunction';
import { showCreateTable } from './showCreateTable';

describe('showCreateTable', () => {
  let dbConnection: DatabaseConnection;
  beforeAll(async () => {
    dbConnection = await getDbConnection({ language: DatabaseLanguage.POSTGRES });
  });
  afterAll(async () => {
    await dbConnection.end();
  });
  it('should be possible to get create statement of table', async () => {
    await provisionShowCreateTableFunction({ dbConnection });
    await dbConnection.query({ sql: 'DROP TABLE IF EXISTS test_table_for_showcreate_on;' });
    await dbConnection.query({
      sql: `
CREATE TABLE test_table_for_showcreate_on (
  id BIGSERIAL PRIMARY KEY,
  name VARCHAR(150),
  level VARCHAR(50) CHECK (level IN ('info', 'warn', 'error'))
)
    `.trim(),
    });
    const result = await showCreateTable({ dbConnection, schema: 'public', table: 'test_table_for_showcreate_on' });
    expect(result).toMatchSnapshot();
  });
});
