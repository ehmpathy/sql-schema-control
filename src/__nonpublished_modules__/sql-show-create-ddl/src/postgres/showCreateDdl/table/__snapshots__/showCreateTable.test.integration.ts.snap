// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showCreateTable should be possible to get create statement of table 1`] = `
"CREATE TABLE public.test_table_for_showcreate_on (
  id bigint NOT NULL DEFAULT nextval('test_table_for_showcreate_on_id_seq'::regclass),
  name character varying(150) NULL,
  level character varying(50) NULL,
  CONSTRAINT test_table_for_showcreate_on_pkey PRIMARY KEY (id),
  CONSTRAINT test_table_for_showcreate_on_level_check CHECK (((level)::text = ANY ((ARRAY['info'::character varying, 'warn'::character varying, 'error'::character varying])::text[])))
);
"
`;
